### PR TITLE
Add configure option to not install ".byte" executables

### PR DESCRIPTION
**Pull request by mshinwell**
In environments where the executables compiled to native code, such as ocamlopt.opt, are always used in preference to the bytecode versions then space can be saved by not installing the latter.  This patch provides a configure option to do such.  It is relatively lightly engineered; in particular, it won't complain if the native code executables aren't themselves being built; but given this is an option for knowledgeable users I think that is reasonable.

### CODE REVIEW

**Comment by shindere** [comment]
At least two things:

1. IMO, having varialbes like NO_INSTALLED_BYTECODE_EXES which must be
false for the feature to be actually enabled is counter-intuitive, hard
to maintain and error-prone. I'd prefer a variable that simply tells
whether the feature is enabled and then installing the files if its
value is true, even if that makes the patch bigger.

2. The name you have chosen for the configure option won't work well
with the autoconf generated script. It should rather be a name like
--disable-bytecode-executables-installation or something similar.

**Comment by ghost** [comment]
Why is it not the default? Is there any point in installing the bytecode versions?

**Comment by mshinwell** [comment]
@shindere I think point 1 is arguable either way.  I take the "feature", that is to say something to be optionally enabled, as the _disabling_ of the installation of bytecode executables.  The common case is to not have the feature.

For point 2, the naming of this option follows the other options that currently exist in the `configure` script, no?

**Comment by xclerc** [comment]
Not very important, but it could be extended to `ocamldoc`
(which, by the way, does not install through `INSTALL_PROG`).

**Comment by shindere** [comment]
Mark Shinwell (2018/05/09 03:27 -0700):
> For point 2, the naming of this option follows the other options that
> currently exist in the `configure` script, no?

Yes, but these will beproblematic, too, so it's perhaps better to give
it the right name from the beginning than to stick to the old convension
and then have to change it later.

For point 1, although one can theoretically argue both ways, I do think
that, from the point of view of the readability of the resulting code,
it would be more readable the other way around.

**Comment by mshinwell** [comment]
Well, I would have thought having one convention which is followed now, then a different convention later (at which point everything changes) was better than two conventions now then one later.  Maybe that's just me though.

As regards @diml 's point: maybe this should be the default, but if it is the default, there should probably be an option to turn it off (in particular for use when developing the native code compiler, which sometimes causes miscompilations of the `.opt` versions).  Also, a discussion about whether this should be the default can be separate from this pull request.

**Comment by gasche** [comment]
New configure options should always come in both a negative or a positive version, except in the rare case where we feel exceedingly sure that the default value will never change in the future -- which is clearly not the case here. So `{enable,disable}`-bytecode-executables as suggested by @shindere, or `{with,no}-bytecode-executables` if you prefer the current partial convention (`{with,no}-{debug,instrumented}-runtime`, `{with,no}-cplugins`, etc.). (Personally I would follow the current convention.)

Besides the command-line settings, I agree with @shindere that a positive name for the internal option (`WITH_BYTECODE_EXECS`) would make more sense and be easier to maintain.

**Comment by shindere** [comment]
Gabriel Scherer (2018/05/09 05:31 -0700):
> New configure options should always come in both a negative or a
> positive version, except in the rare case where we feel exceedingly
> sure that the default value will never change in the future -- which
> is clearly not the case here.

Agreed!

> So `{enable,disable}`-bytecode-executables as suggested by @shindere, or
> `{with,no}-bytecode-executables` if you prefer the current partial
> convention (`{with,no}-{debug,instrumented}-runtime`,
> `{with,no}-cplugins`, etc.). (Personally I would follow the current
> convention.)

I don't mind following the current convention and then changing
everything at once, if that's preferer.

Re:the option name, I think it should include "install". Although it
makes the option name long, I think if "install" is not mentionned users
may infer that it's about building.

> Besides the command-line settings, I agree with @shindere that a
> positive name for the internal option (`WITH_BYTECODE_EXECS`) would
> make more sense and be easier to maintain.

I'd actually suggest INSTALL_BYTECODE_PROGRAMS

**Comment by mshinwell** [comment]
I have pushed a revised version.

**Comment by gasche** [comment]
The configure options are fine (`-do-install` and `-do-not-install` would sound better, but I'm content with `-install` and `-no-install`) and the naming is better. Maybe it would sense to also comment out the actions in the style of `$(LN) ocamlc.byte$(EXE) ocamlc$(EXE)` which create dead links with the current patch?

(You pointed out that you don't try to file if `-no-install-...` is used while native compilation is not available, but maybe we could make sure we don't create dead links in the install directory.)

I used `git grep "LN.*\\.byte.*"` to catch the occurrences of this pattern.

**Comment by mshinwell** [comment]
Not creating the links is probably fine.  @shindere do you have any other comments?

**Comment by mshinwell** [comment]
@gasche It doesn't create bad links now.  Please approve if this is now satisfactory.

**Review by gasche — APPROVED** [review]
Approved -- if the CI passes -- but you should rebase/squash before or when merging.

**Pull request merged**
